rust/worker/src/execution/operators/merge_metadata_results.rs

```diff
@@ -107,289 +113,136 @@ impl Operator<MergeMetadataResultsOperatorInput, MergeMetadataResultsOperatorOut
-                                // This means no compaction has occured, so we can just return whats on the log.
-                                return Ok(MergeMetadataResultsOperatorOutput {
-                                    ids,
-                                    metadata,
-                                    documents,
-                                });
-                            }
-                            RecordSegmentReaderCreationError::BlockfileOpenError(_) => {
-                                error!("Error creating Record Segment: {:?}", e);
-                                return Err(
-                                    MergeMetadataResultsOperatorError::RecordSegmentCreationError(
-                                        *e,
-                                    ),
-                                );
-                            }
-                            RecordSegmentReaderCreationError::InvalidNumberOfFiles => {
-                                error!("Error creating Record Segment: {:?}", e);
-                                return Err(
-                                    MergeMetadataResultsOperatorError::RecordSegmentCreationError(
-                                        *e,
-                                    ),
-                                );
-                            }
-                        }
+                .then_some(log.offset_id)
+            }));
+
+        // Merge the offset ids, assuming the user_offset_ids and filtered_offset_ids are ordered.
+        let merged_oids_holder: Vec<u32>;
+        let merged_offset_ids = match (&input.filtered_offset_ids, &input.user_offset_ids) {
+            (Some(fids), Some(uids)) => {
+                merged_oids_holder = merge_sorted_vecs_conjunction(fids, uids);
+                &merged_oids_holder
+            }
+            (Some(oids), None) | (None, Some(oids)) => oids,
+            // If both filter and user offset ids are None, it suggests user did not specify any filter. We fetch all offset ids using the record segment reader
+            _ => {
+                let mut live_log_offset_ids = mat_records
+                    .iter()
+                    .map(|(log, _)| log.offset_id)
+                    .collect::<Vec<_>>();
+                live_log_offset_ids.sort();
+                merged_oids_holder = match &record_segment_reader {
+                    Some(reader) => {
+                        // TODO: We may not need to have the entire offset ids, given limit and offset.
+                        let compact_offset_ids =
+                            reader.get_all_offset_ids().await.map_err(|e| {
+                                tracing::error!("Error reading record segment: {}", e);
+                                MergeMetadataResultsOperatorError::RecordSegmentReadError
+                            })?;
+                        merge_sorted_vecs_disjunction(&live_log_offset_ids, &compact_offset_ids)
```

Comment:
this seems incorrect. Records that are present on the log in deleted state will be included in the output which is incorrect